cpp/arcticdb/storage/azure/azure_storage.cpp

```diff
@@ -47,6 +47,61 @@ namespace fg = folly::gen;
 
 namespace detail {
 
+// TODO: fix this temporary workaround to read error code. azure-sdk-cpp client sometimes doesn't properly set the error code.
+//  This issue has been raised on the sdk repo. Once fixed, we should no longer need the following function and would just read e.ErrorCode.
+std::string get_error_code(const Azure::Core::RequestFailedException& e) {
+    auto error_code = e.ErrorCode;
+
+    if(error_code.empty() && e.RawResponse ) {
+        auto headers_map = e.RawResponse->GetHeaders();
+        if(headers_map.find("x-ms-error-code") != headers_map.end()) {
+            error_code = headers_map["x-ms-error-code"];
+        }
+    }
+    return error_code;
+}
+
+void raise_azure_exception(const Azure::Core::RequestFailedException& e) {
+    auto error_code = get_error_code(e);
+    auto status_code = e.StatusCode;
+    std::string error_message;
+
+    if(status_code == Azure::Core::Http::HttpStatusCode::NotFound && error_code == AzureErrorCode_to_string(AzureErrorCode::BlobNotFound)) {
+        throw KeyNotFoundException(fmt::format("Key Not Found Error: AzureError#{} {}: {}",
+                                               int(status_code), error_code, e.ReasonPhrase));
+    }
+
+    if(status_code == Azure::Core::Http::HttpStatusCode::Unauthorized || status_code == Azure::Core::Http::HttpStatusCode::Forbidden) {
+        raise<ErrorCode::E_PERMISSION>(fmt::format("Permission error: AzureError#{} {}: {}",
+                                                   int(status_code), error_code, e.ReasonPhrase));
+    }
+
+    if((int) status_code >= 500) {
+        error_message = fmt::format("Unexpected Server Error: AzureError#{} {}: {} {}",
+                                    int(status_code), error_code, e.ReasonPhrase, e.what());
+    }
+    else {
+        error_message = fmt::format("Unexpected Error: AzureError#{} {}: {} {}",
+                                    int(status_code), error_code, e.ReasonPhrase, e.what());
+    }
+
+    raise<ErrorCode::E_UNEXPECTED_AZURE_ERROR>(error_message);
+}
+
+bool is_expected_error_type(const std::string& error_code, Azure::Core::Http::HttpStatusCode status_code) {
+    return (status_code == Azure::Core::Http::HttpStatusCode::NotFound && error_code == AzureErrorCode_to_string(AzureErrorCode::BlobNotFound)) ||
+           (status_code == Azure::Core::Http::HttpStatusCode::NotFound && error_code == AzureErrorCode_to_string(AzureErrorCode::ContainerNotFound));
```

Comment:
` return status_code == Azure::Core::Http::HttpStatusCode::NotFound && (error_code == AzureErrorCode_to_string(AzureErrorCode::BlobNotFound) || AzureErrorCode_to_string(AzureErrorCode::ContainerNotFound));`?

Reply:
Fixed in [4ea3442](https://github.com/man-group/ArcticDB/pull/1344/commits/4ea3442882bfdc4001ff117531804a8ba28a8d08)